src/components/calcite-panel/calcite-panel.tsx

```diff
@@ -180,45 +342,126 @@ export class CalcitePanel {
-    const headerTrailingContentNode = this.renderHeaderTrailingContent();
+  renderMenuButton(): VNode {
+    const { menuOpen, intlOpen, intlClose } = this;
+    const closeLabel = intlClose || TEXT.close;
+    const openLabel = intlOpen || TEXT.open;
 
-    const canDisplayHeader =
-      headerContentNode || headerLeadingContentNode || headerTrailingContentNode;
+    const menuLabel = menuOpen ? closeLabel : openLabel;
+
+    return (
+      <calcite-action
+        class={CSS.menuButton}
+        aria-label={menuLabel}
+        text={menuLabel}
+        ref={(menuButtonEl): HTMLCalciteActionElement => (this.menuButtonEl = menuButtonEl)}
+        onClick={this.toggleMenuOpen}
+        onKeyDown={this.menuButtonKeyDown}
+        icon={ICONS.menu}
+      />
+    );
+  }
 
-    return canDisplayHeader ? (
+  renderMenu(): VNode {
+    const { el } = this;
+    const hasMenuItems = getSlotted(el, SLOTS.headerMenuActions);
+
+    return hasMenuItems ? (
+      <div class={CSS.menuContainer} onKeyDown={this.menuActionsContainerKeyDown}>
+        {this.renderMenuButton()}
+        {this.renderMenuItems()}
+      </div>
+    ) : null;
+  }
+
+  renderHeaderNode(): VNode {
+    const { el } = this;
+    const hasHeaderSlottedContent = getSlotted(el, SLOTS.headerContent);
+    const backButtonNode = this.renderBackButton();
+    const hasStartActions = getSlotted(el, SLOTS.headerActionsStart);
+    const headerContentNode = hasHeaderSlottedContent
+      ? this.renderHeaderSlottedContent()
+      : this.renderHeaderContent();
+    const endActionsNode = this.renderHeaderEndContent();
+    const headerMenuNode = this.renderMenu();
+
+    return hasStartActions || headerContentNode || endActionsNode ? (
       <header class={CSS.header}>
-        {headerLeadingContentNode}
+        {backButtonNode}
+        <slot name={SLOTS.headerActionsStart} />
         {headerContentNode}
-        {headerTrailingContentNode}
+        {endActionsNode}
+        {headerMenuNode}
       </header>
     ) : null;
   }
 
-  renderFooter(): VNode {
+  /**
+   * Allows user to override the entire footer node.
+   */
+  renderFooterSlottedContent(): VNode {
     const { el } = this;
 
-    const hasFooter = getSlotted(el, SLOTS.footer);
+    const hasFooterSlottedContent = getSlotted(el, SLOTS.footer);
 
-    return hasFooter ? (
+    return hasFooterSlottedContent ? (
       <footer class={CSS.footer}>
         <slot name={SLOTS.footer} />
       </footer>
     ) : null;
   }
 
+  renderFooterActions(): VNode {
+    const { el } = this;
+
+    const hasFooterActions = getSlotted(el, SLOTS.footerActions);
+    
+    return hasFooterActions ? (
+      <footer class={CSS.footer}>
+        <slot name={SLOTS.footerActions}></slot>
+      </footer>
+    ) : null;
+  }
+
+  renderFooter(): VNode {
+    const footerContentNode = this.renderFooterSlottedContent()
+    const footerActionsNode = this.renderFooterActions();
+
+    return footerContentNode || footerActionsNode;
```

Comment:
Lets just simplify it to `return this.renderFooterSlottedContent() || this.renderFooterActions();`

Reply:
Copy that.

Reply:
Updated. Let me know if I went too far on simplifying.